# Sentiment_Finbert

| 항목 | 내용 |
|------|------|
| 모델 | FinBERT (`ProsusAI/finbert`) |
| 입력 데이터 | Reuters 뉴스 headline |
| 분석 단위 | 문장(sentence) 단위 감성분석 |
| 감성 클래스 | positive · neutral · negative |
| 활용 목적 | 유가 관련 뉴스 감성의 정량화 |

--------------------------------------------------

> **참고** 
> Google News RSS의 `summary`는 headline과 중복되는 경우가 많아  감성분석 입력에는 headline만 사용하였습니다.  

In [26]:
from pathlib import Path
import pandas as pd

DATA_DIR = Path("../data/Crawling_Final")

CATEGORIES = ["supply", "demand", "geopolitical", "financial"]

MERGED_FILES = {
    cat: DATA_DIR / f"merged_Oil_{cat.capitalize()}_Final.csv"
    for cat in CATEGORIES
}

category_news_dfs = {}

for cat, file_path in MERGED_FILES.items():
    if not file_path.exists():
        print(f"[{cat}] 파일 없음 → {file_path}")
        continue

    df = pd.read_csv(file_path)

    # 날짜 형식 정리
    df["date"] = pd.to_datetime(df["date"], errors="coerce")

    # headline 결측 제거
    df = df[df["headline"].notna()].copy()

    # category 컬럼 보정
    df["category"] = cat

    category_news_dfs[cat] = df

    print(f"[{cat}] 로드 완료: {len(df):,}건")
    print(f"  기간: {df['date'].min().date()} ~ {df['date'].max().date()}")

[supply] 로드 완료: 16,903건
  기간: 2007-08-09 ~ 2026-03-31
[demand] 로드 완료: 25,752건
  기간: 2007-08-09 ~ 2026-03-31
[geopolitical] 로드 완료: 46,505건
  기간: 2007-08-02 ~ 2026-03-31
[financial] 로드 완료: 21,989건
  기간: 2007-08-02 ~ 2026-03-31


In [27]:
for cat, df in category_news_dfs.items():
    print(f"\n=== [{cat}] ===")
    display(df.head())
    print(df.columns.tolist())


=== [supply] ===


,date,headline,summary,url,source,category,keyword,oil_relevance,is_noise,range_tag
0,2007-08-09,Halliburton distances itself from KBR,Halliburton distances itself from KBR Reuters,https://news.google.com/rss/articles/CBMimAFBV...,Reuters,supply,oil drilling activity,0,True,08.07-12.10
1,2007-08-09,"FTSE extends gains on UK, U.S. data; PPI in focus","FTSE extends gains on UK, U.S. data; PPI in fo...",https://news.google.com/rss/articles/CBMiowFBV...,Reuters,supply,oil inventory,0,True,08.07-12.10
2,2007-08-09,"Iran, banks weigh on European shares; M&A feat...","Iran, banks weigh on European shares; M&A feat...",https://news.google.com/rss/articles/CBMipAFBV...,Reuters,supply,oil inventory,-1,True,08.07-12.10
3,2007-08-09,"Oil eases, near $62, on summer fuel concerns","Oil eases, near $62, on summer fuel concerns R...",https://news.google.com/rss/articles/CBMimgFBV...,Reuters,supply,oil inventory,2,False,08.07-12.10
4,2007-08-09,US shares rise as Exxon's gain drowns out Bern...,US shares rise as Exxon's gain drowns out Bern...,https://news.google.com/rss/articles/CBMipwFBV...,Reuters,supply,oil inventory,-1,True,08.07-12.10


['date', 'headline', 'summary', 'url', 'source', 'category', 'keyword', 'oil_relevance', 'is_noise', 'range_tag']

=== [demand] ===


,date,headline,summary,url,source,category,keyword,oil_relevance,is_noise,range_tag
0,2007-08-09,"Vilma Espin, first lady of Cuban revolution, dies","Vilma Espin, first lady of Cuban revolution, d...",https://news.google.com/rss/articles/CBMiogFBV...,Reuters,demand,oil consumption,0,True,08.07-12.10
1,2007-08-09,India resumes mango exports to United States,India resumes mango exports to United States R...,https://news.google.com/rss/articles/CBMioAFBV...,Reuters,demand,global oil demand,0,True,08.07-12.10
2,2007-08-09,Mamas & Papas survivor says she's over singing,Mamas & Papas survivor says she's over singing...,https://news.google.com/rss/articles/CBMiogFBV...,Reuters,demand,diesel demand,0,True,08.07-12.10
3,2007-08-09,Intel to treble Vietnam investment to $1 bln: ...,Intel to treble Vietnam investment to $1 bln: ...,https://news.google.com/rss/articles/CBMiqwFBV...,Reuters,demand,global oil demand,0,True,08.07-12.10
4,2007-08-09,"Oil eases, near $62, on summer fuel concerns","Oil eases, near $62, on summer fuel concerns R...",https://news.google.com/rss/articles/CBMimgFBV...,Reuters,demand,oil gasoline demand,2,False,08.07-12.10


['date', 'headline', 'summary', 'url', 'source', 'category', 'keyword', 'oil_relevance', 'is_noise', 'range_tag']

=== [geopolitical] ===


,date,headline,summary,url,source,category,keyword,oil_relevance,is_noise,range_tag
0,2007-08-02,Moody's Commodities Index,Moody's Commodities Index Reuters,https://news.google.com/rss/articles/CBMihgFBV...,Reuters,geopolitical,middle east crude oil,0,True,08.07-12.10
1,2007-08-09,Quebecers said perfect gene pool for medical r...,Quebecers said perfect gene pool for medical r...,https://news.google.com/rss/articles/CBMi0gFBV...,Reuters,geopolitical,middle east crude oil,0,True,08.07-12.10
2,2007-08-09,Russian accuses MI6 of trying to recruit him,Russian accuses MI6 of trying to recruit him R...,https://news.google.com/rss/articles/CBMiogFBV...,Reuters,geopolitical,russia ukraine oil,2,False,08.07-12.10
3,2007-08-09,"Singapore's Cosco 2006 profit up 28%, upbeat o...","Singapore's Cosco 2006 profit up 28%, upbeat o...",https://news.google.com/rss/articles/CBMiqgFBV...,Reuters,geopolitical,russia ukraine oil,0,True,08.07-12.10
4,2007-08-09,Russia's richest man Abramovich splits from wife,Russia's richest man Abramovich splits from wi...,https://news.google.com/rss/articles/CBMiogFBV...,Reuters,geopolitical,russia ukraine oil,2,False,08.07-12.10


['date', 'headline', 'summary', 'url', 'source', 'category', 'keyword', 'oil_relevance', 'is_noise', 'range_tag']

=== [financial] ===


,date,headline,summary,url,source,category,keyword,oil_relevance,is_noise,range_tag
0,2007-08-02,Moody's Commodities Index,Moody's Commodities Index Reuters,https://news.google.com/rss/articles/CBMihgFBV...,Reuters,financial,dollar crude oil price,0,True,08.07-12.10
1,2007-08-09,Abuse of prescribed slimming drug seen in HK,Abuse of prescribed slimming drug seen in HK R...,https://news.google.com/rss/articles/CBMixgFBV...,Reuters,financial,crude oil futures,0,True,08.07-12.10
2,2007-08-09,Shepherds Select scandal -- the fallout,Shepherds Select scandal -- the fallout Reuters,https://news.google.com/rss/articles/CBMilgFBV...,Reuters,financial,dollar crude oil price,0,True,08.07-12.10
3,2007-08-09,Red Robin ex-CEO settles SEC fraud charges ove...,Red Robin ex-CEO settles SEC fraud charges ove...,https://news.google.com/rss/articles/CBMiqwFBV...,Reuters,financial,dollar crude oil price,0,True,08.07-12.10
4,2007-08-09,Indian shares stall before record high;Karachi...,Indian shares stall before record high;Karachi...,https://news.google.com/rss/articles/CBMiqgFBV...,Reuters,financial,dollar crude oil price,-1,True,08.07-12.10


['date', 'headline', 'summary', 'url', 'source', 'category', 'keyword', 'oil_relevance', 'is_noise', 'range_tag']


## 1. oil_relevance 필터링

In [28]:
RELEVANCE_THRESHOLD = 1

filtered_category_news_dfs = {}

for cat, df in category_news_dfs.items():
    before = len(df)

    df = df.copy()

    # oil_relevance 숫자형 보장
    # 크롤링 CSV 재로드 시 문자열/헤더 행이 섞이면 coerce로 NaN 처리
    df["oil_relevance"] = pd.to_numeric(df["oil_relevance"], errors="coerce")

    # oil_relevance 기준으로 원유 관련성이 있는 기사만 유지
    df_filtered = df[
        df["oil_relevance"] >= RELEVANCE_THRESHOLD
    ].copy()

    after = len(df_filtered)

    filtered_category_news_dfs[cat] = df_filtered

    print(f"[{cat}] relevance filtering")
    print(f"  before: {before:,}건")
    print(f"  after : {after:,}건")
    print(f"  removed: {before - after:,}건 ({(before - after) / before * 100:.1f}%)")

[supply] relevance filtering
  before: 16,903건
  after : 8,925건
  removed: 7,978건 (47.2%)
[demand] relevance filtering
  before: 25,752건
  after : 8,079건
  removed: 17,673건 (68.6%)
[geopolitical] relevance filtering
  before: 46,505건
  after : 16,758건
  removed: 29,747건 (64.0%)
[financial] relevance filtering
  before: 21,989건
  after : 8,367건
  removed: 13,622건 (61.9%)


In [29]:
filter_summary = []

for cat in category_news_dfs:
    before = len(category_news_dfs[cat])
    after = len(filtered_category_news_dfs[cat])

    filter_summary.append({
        "category": cat,
        "before": before,
        "after": after,
        "removed": before - after,
        "remaining_rate": after / before if before > 0 else 0,
        "removed_rate": (before - after) / before if before > 0 else 0,
    })

filter_summary_df = pd.DataFrame(filter_summary)

display(
    filter_summary_df.assign(
        remaining_rate=lambda x: (x["remaining_rate"] * 100).round(1),
        removed_rate=lambda x: (x["removed_rate"] * 100).round(1),
    )
)


,category,before,after,removed,remaining_rate,removed_rate
0,supply,16903,8925,7978,52.8,47.2
1,demand,25752,8079,17673,31.4,68.6
2,geopolitical,46505,16758,29747,36.0,64.0
3,financial,21989,8367,13622,38.1,61.9


In [30]:
# 필터링 후 키워드별 기사 수 확인
for cat, df in filtered_category_news_dfs.items():
    print(f"\n=== [{cat}] 필터링 후 키워드별 기사 수 ===")
    print(df["keyword"].value_counts().to_string())


=== [supply] 필터링 후 키워드별 기사 수 ===
keyword
US shale oil production        2960
OPEC production                2670
crude oil exports              1834
oil inventory                   962
crude oil supply disruption     291
oil drilling activity           144
OPEC quota                       38
OPEC spare capacity              26

=== [demand] 필터링 후 키워드별 기사 수 ===
keyword
oil gasoline demand           3288
oil consumption               1751
global oil demand             1392
jet fuel demand               1070
diesel demand                  515
crude oil demand recession      63

=== [geopolitical] 필터링 후 키워드별 기사 수 ===
keyword
russia ukraine oil                 4006
middle east crude oil              3078
iran crude oil sanctions           2800
geopolitical risk oil              1327
oil price war                      1286
strategic petroleum reserve oil    1235
crude oil supply disruption         852
russia crude oil exports            784
oil supply conflict                 742
crude oil 

## 2. 카테고리별 FinBERT 실행

#### 2-1. FinBERT 입력 데이터 생성

In [7]:
finbert_input_dfs = {}

for cat, df in filtered_category_news_dfs.items():
    df_input = df.copy()

    # FinBERT 입력은 headline만 사용
    df_input["finbert_text"] = (
        df_input["headline"]
        .fillna("")
        .astype(str)
        .str.strip()
    )

    # 빈 headline 제거
    before = len(df_input)
    df_input = df_input[df_input["finbert_text"] != ""].copy()
    after = len(df_input)

    finbert_input_dfs[cat] = df_input

    print(f"[{cat}] FinBERT input 준비")
    print(f"  before: {before:,}건")
    print(f"  after : {after:,}건")
    print(f"  removed empty headline: {before - after:,}건")

[supply] FinBERT input 준비
  before: 8,925건
  after : 8,925건
  removed empty headline: 0건
[demand] FinBERT input 준비
  before: 8,079건
  after : 8,079건
  removed empty headline: 0건
[geopolitical] FinBERT input 준비
  before: 16,758건
  after : 16,758건
  removed empty headline: 0건
[financial] FinBERT input 준비
  before: 8,367건
  after : 8,367건
  removed empty headline: 0건


** [Headline만 사용한 이유]

: Reuters 웹사이트는 DataDome 봇 차단으로 직접 크롤링이 불가하여 Google News RSS를 통해 수집하였다. 이때, RSS로 수집한 summary는 대부분 headline과 동일한 내용을 반환하므로, 감성분석 및 관련성 평가는 headline 텍스트만을 대상으로 수행하였다.

In [8]:
# summary == headline + " Reuters"
from pathlib import Path
import pandas as pd

DATA_DIR = Path("../data/Crawling_Final")
files = sorted(DATA_DIR.glob("merged_Oil_*_Final.csv"))

df = pd.concat([pd.read_csv(f) for f in files], ignore_index=True)

headline = df["headline"].fillna("").astype(str).str.strip()
summary = df["summary"].fillna("").astype(str).str.strip()

check = summary.eq(headline + " Reuters")

print(f"전체 건수: {len(df):,}")
print(f"headline + ' Reuters'와 summary가 일치한 건수: {check.sum():,}")
print(f"불일치 건수: {(~check).sum():,}")
print(f"일치 비율: {check.mean():.2%}")

전체 건수: 111,149
headline + ' Reuters'와 summary가 일치한 건수: 111,130
불일치 건수: 19
일치 비율: 99.98%


In [9]:
# 각 카테고리별 FinBERT 입력 데이터 샘플
for cat, df in finbert_input_dfs.items():
    print(f"\n=== [{cat}] FinBERT input sample ===")
    display(
        df[["date", "category", "keyword", "headline", "finbert_text", "oil_relevance"]].head()
    )


=== [supply] FinBERT input sample ===


,date,category,keyword,headline,finbert_text,oil_relevance
3,2007-08-09,supply,oil inventory,"Oil eases, near $62, on summer fuel concerns","Oil eases, near $62, on summer fuel concerns",2
5,2007-08-09,supply,oil inventory,"Cisco lifts Nasdaq, S&P; oil takes shine off Dow","Cisco lifts Nasdaq, S&P; oil takes shine off Dow",1
15,2007-08-09,supply,OPEC production,Venezuela seizes operations from big oil compa...,Venezuela seizes operations from big oil compa...,1
17,2007-08-09,supply,OPEC production,Oil above $52 as U.S. cold snap supports buying,Oil above $52 as U.S. cold snap supports buying,1
18,2007-08-09,supply,OPEC production,FACTBOX-Nigeria's oil production outages,FACTBOX-Nigeria's oil production outages,3



=== [demand] FinBERT input sample ===


,date,category,keyword,headline,finbert_text,oil_relevance
4,2007-08-09,demand,oil gasoline demand,"Oil eases, near $62, on summer fuel concerns","Oil eases, near $62, on summer fuel concerns",2
5,2007-08-09,demand,oil gasoline demand,Oil above $52 as U.S. cold snap supports buying,Oil above $52 as U.S. cold snap supports buying,2
6,2007-08-09,demand,oil gasoline demand,"U.S. oil pipelines won't carry biodiesel, yet","U.S. oil pipelines won't carry biodiesel, yet",3
15,2007-08-09,demand,global oil demand,McDonald's to recycle cooking oil for fuel,McDonald's to recycle cooking oil for fuel,2
32,2007-08-09,demand,global oil demand,Scented oils linked to breast growth in boys,Scented oils linked to breast growth in boys,2



=== [geopolitical] FinBERT input sample ===


,date,category,keyword,headline,finbert_text,oil_relevance
2,2007-08-09,geopolitical,russia ukraine oil,Russian accuses MI6 of trying to recruit him,Russian accuses MI6 of trying to recruit him,2
4,2007-08-09,geopolitical,russia ukraine oil,Russia's richest man Abramovich splits from wife,Russia's richest man Abramovich splits from wife,2
5,2007-08-09,geopolitical,russia ukraine oil,Zimbabwe hunts spirit medium over false fuel c...,Zimbabwe hunts spirit medium over false fuel c...,1
6,2007-08-09,geopolitical,russia ukraine oil,Russia says mysterious yellow snow not toxic,Russia says mysterious yellow snow not toxic,2
7,2007-08-09,geopolitical,russia ukraine oil,Vodka and isolation: welcome to rural Russia,Vodka and isolation: welcome to rural Russia,2



=== [financial] FinBERT input sample ===


,date,category,keyword,headline,finbert_text,oil_relevance
5,2007-08-09,financial,Brent crude,"Oil eases, near $62, on summer fuel concerns","Oil eases, near $62, on summer fuel concerns",1
6,2007-08-09,financial,Brent crude,Oil above $52 as U.S. cold snap supports buying,Oil above $52 as U.S. cold snap supports buying,1
32,2007-08-09,financial,crude oil futures,Merrill sells off assets from Bear hedge funds,Merrill sells off assets from Bear hedge funds,1
36,2007-08-09,financial,crude oil futures,"Cisco lifts Nasdaq, S&P; oil takes shine off Dow","Cisco lifts Nasdaq, S&P; oil takes shine off Dow",1
37,2007-08-09,financial,crude oil futures,"TREASURIES-Bonds fall on mortgage selling, oil...","TREASURIES-Bonds fall on mortgage selling, oil...",2


#### 2-2. FinBERT 모델 로드 및 함수 정의

In [10]:
# import sys
# !{sys.executable} -m pip install torch --index-url https://download.pytorch.org/whl/cpu
# !{sys.executable} -m pip install transformers tqdm

In [11]:
import torch
from transformers import AutoTokenizer, AutoModelForSequenceClassification
from torch.nn.functional import softmax
from tqdm.auto import tqdm

c:\Users\이은서\AppData\Local\Programs\Python\Python313\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [12]:
MODEL_NAME = "ProsusAI/finbert"

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(f"device: {device}")

tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME)
model = AutoModelForSequenceClassification.from_pretrained(MODEL_NAME).to(device)
model.eval()

id2label = model.config.id2label
print(id2label)

device: cpu


Loading weights: 100%|██████████| 201/201 [00:00<00:00, 70791.43it/s]

{0: 'positive', 1: 'negative', 2: 'neutral'}


In [13]:
def run_finbert(texts, batch_size=32):
    results = []

    for i in tqdm(range(0, len(texts), batch_size)):
        batch_texts = texts[i:i + batch_size]

        inputs = tokenizer(
            batch_texts,
            padding=True,
            truncation=True,
            max_length=128,
            return_tensors="pt"
        ).to(device)

        with torch.no_grad():
            outputs = model(**inputs)
            probs = softmax(outputs.logits, dim=1).cpu().numpy()

        for prob in probs:
            label_probs = {
                id2label[idx].lower(): float(prob[idx])
                for idx in range(len(prob))
            }

            positive_prob = label_probs.get("positive", 0.0)
            neutral_prob = label_probs.get("neutral", 0.0)
            negative_prob = label_probs.get("negative", 0.0)

            sentiment_label = max(label_probs, key=label_probs.get)
            sentiment_score = positive_prob - negative_prob

            results.append({
                "positive_prob": positive_prob,
                "neutral_prob": neutral_prob,
                "negative_prob": negative_prob,
                "sentiment_label": sentiment_label,
                "sentiment_score": sentiment_score,
            })

    return pd.DataFrame(results)


In [14]:
### 테스트용
test_texts = finbert_input_dfs["supply"]["finbert_text"].head(5).tolist()
test_result = run_finbert(test_texts, batch_size=5)
display(test_result)

100%|██████████| 1/1 [00:00<00:00, 12.02it/s]


,positive_prob,neutral_prob,negative_prob,sentiment_label,sentiment_score
0,0.707491,0.036692,0.255816,positive,0.451675
1,0.586146,0.189505,0.224349,positive,0.361797
2,0.019584,0.234117,0.746299,negative,-0.726715
3,0.851306,0.074140,0.074554,positive,0.776753
4,0.009875,0.029424,0.960701,negative,-0.950827


#### 2-3. 카테고리별 FinBERT 실행 및 저장

In [19]:
SENTIMENT_DIR = Path("../data/Sentiment_raw")
SENTIMENT_DIR.mkdir(parents=True, exist_ok=True)

finbert_result_dfs = {}

for cat, df in finbert_input_dfs.items():
    print(f"\n=== [{cat}] FinBERT 감성분석 시작: {len(df):,}건 ===")

    texts = df["finbert_text"].tolist()
    sentiment_df = run_finbert(texts, batch_size=32)

    result_df = pd.concat(
        [df.reset_index(drop=True), sentiment_df],
        axis=1
    )

    save_path = SENTIMENT_DIR / f"oil_{cat}_with_finbert.csv"
    result_df.to_csv(save_path, index=False, encoding="utf-8-sig")

    finbert_result_dfs[cat] = result_df

    print(f"[{cat}] 저장 완료 → {save_path}")



=== [supply] FinBERT 감성분석 시작: 8,925건 ===


100%|██████████| 279/279 [01:04<00:00,  4.30it/s]


[supply] 저장 완료 → ..\data\Sentiment_raw\oil_supply_with_finbert.csv

=== [demand] FinBERT 감성분석 시작: 8,079건 ===


100%|██████████| 253/253 [01:42<00:00,  2.48it/s]


[demand] 저장 완료 → ..\data\Sentiment_raw\oil_demand_with_finbert.csv

=== [geopolitical] FinBERT 감성분석 시작: 16,758건 ===


100%|██████████| 524/524 [02:01<00:00,  4.33it/s]


[geopolitical] 저장 완료 → ..\data\Sentiment_raw\oil_geopolitical_with_finbert.csv

=== [financial] FinBERT 감성분석 시작: 8,367건 ===


100%|██████████| 262/262 [01:01<00:00,  4.24it/s]

[financial] 저장 완료 → ..\data\Sentiment_raw\oil_financial_with_finbert.csv


In [15]:
### FinBERT 결과 확인

for cat, df in finbert_result_dfs.items():
    print(f"\n=== [{cat}] sentiment label 분포 ===")
    print(df["sentiment_label"].value_counts(normalize=True).round(3))
    display(df[[
        "date", "category", "keyword", "headline",
        "positive_prob", "neutral_prob", "negative_prob",
        "sentiment_label", "sentiment_score"
    ]].head())


NameError: name 'finbert_result_dfs' is not defined

## 3. 카테고리별 일별 aggregation

In [ ]:
DAILY_DIR = Path("../data/Sentiment_Final")
DAILY_DIR.mkdir(parents=True, exist_ok=True)

daily_sentiment_dfs = {}

for cat, df in finbert_result_dfs.items():
    df_daily = df.copy()

    # 날짜 형식 정리
    df_daily["date"] = pd.to_datetime(df_daily["date"], errors="coerce")
    df_daily = df_daily[df_daily["date"].notna()].copy()

    # 날짜 단위로 변환
    df_daily["date"] = df_daily["date"].dt.date

    # 일별 aggregation
    # oil_relevance는 진단/검토용 평균값으로 저장하며,
    # sentiment 평균 계산의 가중치로는 사용하지 않음
    daily = (
        df_daily.groupby("date")
        .agg(
            **{
                f"sentiment_{cat}": ("sentiment_score", "mean"),
                f"oil_relevance_{cat}": ("oil_relevance", "mean"),
                f"news_count_{cat}": ("headline", "count"),
                f"positive_prob_{cat}": ("positive_prob", "mean"),
                f"neutral_prob_{cat}": ("neutral_prob", "mean"),
                f"negative_prob_{cat}": ("negative_prob", "mean"),
            }
        )
        .reset_index()
    )

    daily["date"] = pd.to_datetime(daily["date"])

    save_path = DAILY_DIR / f"daily_sentiment_{cat}.csv"
    daily.to_csv(save_path, index=False, encoding="utf-8-sig")

    daily_sentiment_dfs[cat] = daily

    print(f"[{cat}] daily sentiment 저장 완료 → {save_path}")
    print(f"  기간: {daily['date'].min().date()} ~ {daily['date'].max().date()}")
    print(f"  날짜 수: {len(daily):,}")


[supply] daily sentiment 저장 완료 → ..\data\Sentiment_Final\daily_sentiment_supply.csv
  기간: 2007-08-09 ~ 2026-03-31
  날짜 수: 2,472
[demand] daily sentiment 저장 완료 → ..\data\Sentiment_Final\daily_sentiment_demand.csv
  기간: 2007-08-09 ~ 2026-03-31
  날짜 수: 2,532
[geopolitical] daily sentiment 저장 완료 → ..\data\Sentiment_Final\daily_sentiment_geopolitical.csv
  기간: 2007-08-09 ~ 2026-03-31
  날짜 수: 4,460
[financial] daily sentiment 저장 완료 → ..\data\Sentiment_Final\daily_sentiment_financial.csv
  기간: 2007-08-09 ~ 2026-03-31
  날짜 수: 2,997


[가중치]

- 키워드 빈도수에 대한 가중치

=> 날짜별로 sentiment_score를 기사 단위 관측치를 기준으로 평균하므로, 특정 키워드에서 더 많은 기사가 수집된 경우 해당 키워드의 sentiment가 일별 지표에 더 크게 반영된다. 따라서 키워드 빈도 기반 가중 효과가 내재되어 있다. 

- Oil_relevance에 대한 가중치

=> 예를 들어 oil_relevance=3이 oil_relevance=1보다 정확히 3배 더 원유시장 관련성이 높다고 보기는 어려운데, 가중평균에 넣으면 실제로는 3배 가중치를 주는 효과가 생겨 가중치를 적용하지 않았다. 

In [25]:
### 일별 뉴스 커버리지 확인

for cat, df in filtered_category_news_dfs.items():
    daily_count = (
        df.assign(date=pd.to_datetime(df["date"]).dt.date)
        .groupby("date")
        .size()
    )

    start_date = pd.to_datetime(df["date"]).min()
    end_date = pd.to_datetime(df["date"]).max()
    total_days = (end_date - start_date).days + 1

    print(f"\n=== [{cat}] ===")
    print(f"뉴스 있는 날짜 수: {len(daily_count):,}")
    print(f"총 기간 달력일 대비 비율: {len(daily_count) / total_days * 100:.1f}%")
    print(f"평균 일별 기사 수: {daily_count.mean():.2f}")
    print(f"최소 일별 기사 수: {daily_count.min()}")
    print(f"최대 일별 기사 수: {daily_count.max()}")


=== [supply] ===
뉴스 있는 날짜 수: 2,472
총 기간 달력일 대비 비율: 36.3%
평균 일별 기사 수: 3.61
최소 일별 기사 수: 1
최대 일별 기사 수: 23

=== [demand] ===
뉴스 있는 날짜 수: 2,532
총 기간 달력일 대비 비율: 37.2%
평균 일별 기사 수: 3.19
최소 일별 기사 수: 1
최대 일별 기사 수: 19

=== [geopolitical] ===
뉴스 있는 날짜 수: 4,460
총 기간 달력일 대비 비율: 65.5%
평균 일별 기사 수: 3.76
최소 일별 기사 수: 1
최대 일별 기사 수: 52

=== [financial] ===
뉴스 있는 날짜 수: 2,997
총 기간 달력일 대비 비율: 44.0%
평균 일별 기사 수: 2.79
최소 일별 기사 수: 1
최대 일별 기사 수: 20


## 4. 데이터셋 일 단위 확장

In [2]:
import pandas as pd
import numpy as np
from pathlib import Path

# 경로 설정
SENTIMENT_DIR = Path("../data/Sentiment_Final")

categories = ["demand", "supply", "geopolitical", "financial"]

sentiment_files = {
    cat: SENTIMENT_DIR / f"daily_sentiment_{cat}.csv"
    for cat in categories
}

In [3]:
# 카테고리별 Sentiment 데이터 로드
daily_sentiment_dfs = {}

for cat, path in sentiment_files.items():
    df = pd.read_csv(path)
    
    df["date"] = pd.to_datetime(df["date"])
    df = df.sort_values("date").reset_index(drop=True)
    
    daily_sentiment_dfs[cat] = df
    
    print(f"[{cat}] loaded")
    print(f"  path : {path}")
    print(f"  shape: {df.shape}")
    print(f"  date : {df['date'].min().date()} ~ {df['date'].max().date()}")
    print()

[demand] loaded
  path : ..\data\Sentiment_Final\daily_sentiment_demand.csv
  shape: (2532, 7)
  date : 2007-08-09 ~ 2026-03-31

[supply] loaded
  path : ..\data\Sentiment_Final\daily_sentiment_supply.csv
  shape: (2472, 7)
  date : 2007-08-09 ~ 2026-03-31

[geopolitical] loaded
  path : ..\data\Sentiment_Final\daily_sentiment_geopolitical.csv
  shape: (4460, 7)
  date : 2007-08-09 ~ 2026-03-31

[financial] loaded
  path : ..\data\Sentiment_Final\daily_sentiment_financial.csv
  shape: (2997, 7)
  date : 2007-08-09 ~ 2026-03-31



In [4]:
# 카테고리별 시작일 / 마지막일 동일 여부 확인
date_range_summary = []

for cat, df in daily_sentiment_dfs.items():
    date_range_summary.append({
        "category": cat,
        "start_date": df["date"].min(),
        "end_date": df["date"].max(),
        "n_rows": len(df),
        "n_news_days": df["date"].nunique(),
        "total_news_count": df[f"news_count_{cat}"].sum()
    })

date_range_summary = pd.DataFrame(date_range_summary)

display(date_range_summary)

,category,start_date,end_date,n_rows,n_news_days,total_news_count
0,demand,2007-08-09,2026-03-31,2532,2532,8079
1,supply,2007-08-09,2026-03-31,2472,2472,8925
2,geopolitical,2007-08-09,2026-03-31,4460,4460,16758
3,financial,2007-08-09,2026-03-31,2997,2997,8367


In [5]:
# 공통 날짜표 생성(일별)
calendar_start = min(df["date"].min() for df in daily_sentiment_dfs.values())
calendar_end = max(df["date"].max() for df in daily_sentiment_dfs.values())

calendar_df = pd.DataFrame({
    "date": pd.date_range(start=calendar_start, end=calendar_end, freq="D")
})

print("[Common calendar]")
print(f"start: {calendar_start.date()}")
print(f"end  : {calendar_end.date()}")
print(f"days : {len(calendar_df):,}")

[Common calendar]
start: 2007-08-09
end  : 2026-03-31
days : 6,810


In [6]:
# 카테고리별 캘린더 확장
# - news_count는 0으로 채움
# - sentiment는 NaN 유지

calendar_sentiment_dfs = {}

for cat, df in daily_sentiment_dfs.items():
    expanded = calendar_df.merge(df, on="date", how="left")
    
    # 뉴스 없는 날짜 news_count: 0
    expanded[f"news_count_{cat}"] = expanded[f"news_count_{cat}"].fillna(0).astype(int)
    
    # 이진변수(뉴스 없는 날짜(no_news):1, 뉴스 있는 날짜:0)
    expanded[f"no_news_{cat}"] = (expanded[f"news_count_{cat}"] == 0).astype(int)
    
    # sentiment는 NaN 유지
    # oil_relevance, positive_prob 등도 일단 NaN 유지
    
    calendar_sentiment_dfs[cat] = expanded
    
    print(f"[{cat}] calendar expanded")
    print(f"  shape: {expanded.shape}") # 행 수: 6810
    print(f"  date : {expanded['date'].min().date()} ~ {expanded['date'].max().date()}")
    print(f"  no news days: {expanded[f'no_news_{cat}'].sum():,}")
    print()

[demand] calendar expanded
  shape: (6810, 8)
  date : 2007-08-09 ~ 2026-03-31
  no news days: 4,278

[supply] calendar expanded
  shape: (6810, 8)
  date : 2007-08-09 ~ 2026-03-31
  no news days: 4,338

[geopolitical] calendar expanded
  shape: (6810, 8)
  date : 2007-08-09 ~ 2026-03-31
  no news days: 2,350

[financial] calendar expanded
  shape: (6810, 8)
  date : 2007-08-09 ~ 2026-03-31
  no news days: 3,813



In [7]:
# 확장 결과 확인
expanded_summary = []

for cat, df in calendar_sentiment_dfs.items():
    expanded_summary.append({
        "category": cat,
        "calendar_days": len(df),
        "news_days": (df[f"news_count_{cat}"] > 0).sum(),
        "no_news_days": (df[f"news_count_{cat}"] == 0).sum(),
        "coverage_ratio": (df[f"news_count_{cat}"] > 0).mean(),
        "total_news_count": df[f"news_count_{cat}"].sum(),
        "sentiment_na_count": df[f"sentiment_{cat}"].isna().sum()
    })

expanded_summary = pd.DataFrame(expanded_summary)

display(expanded_summary)

,category,calendar_days,news_days,no_news_days,coverage_ratio,total_news_count,sentiment_na_count
0,demand,6810,2532,4278,0.371806,8079,4278
1,supply,6810,2472,4338,0.362996,8925,4338
2,geopolitical,6810,4460,2350,0.654919,16758,2350
3,financial,6810,2997,3813,0.440088,8367,3813


[해석]
- 기간: 2007-08-09 ~ 2026-03-31 (총 6,810일)
- 뉴스가 없는 날짜는 `news_count = 0`, `sentiment = NaN`으로 처리하였다. 
- 확장 후에도 기존 일별 커버리지와 전체 기간 대비 커버리지 비율이 동일함을 확인하였다. 

## 5. 커버리지 확인

#### 5-1. 카테고리별 뉴스 빈도수

In [8]:
category_news_count = []

for cat, df in calendar_sentiment_dfs.items():
    news_col = f"news_count_{cat}"
    
    category_news_count.append({
        "category": cat,
        "news_days": (df[news_col] > 0).sum(),
        "total_news_count": df[news_col].sum(),
        "avg_news_count_news_days": df.loc[df[news_col] > 0, news_col].mean(),
        "min_news_count_news_days": df.loc[df[news_col] > 0, news_col].min(),
        "max_news_count_news_days": df.loc[df[news_col] > 0, news_col].max()
    })

category_news_count = pd.DataFrame(category_news_count)

display(category_news_count)

,category,news_days,total_news_count,avg_news_count_news_days,min_news_count_news_days,max_news_count_news_days
0,demand,2532,8079,3.190758,1,19
1,supply,2472,8925,3.610437,1,23
2,geopolitical,4460,16758,3.757399,1,52
3,financial,2997,8367,2.791792,1,20


[해석]

- 카테고리별 뉴스 빈도와 일별 기사 수에 차이가 있어, 감성점수의 안정성이 카테고리별로 다를 수 있다.
- 따라서 단순 일별 감성점수뿐 아니라 7일 평균 감성과 `news_count`를 함께 고려한다.

#### 5-2. 연도별/주별 커버리지

In [9]:
# 연도별 커버리지
yearly_coverage_list = []

for cat, df in calendar_sentiment_dfs.items():
    tmp = df.copy()
    news_col = f"news_count_{cat}"
    
    # 연도 추출
    tmp["year"] = tmp["date"].dt.year
    
    yearly = (
        tmp.groupby("year")
        .agg(
            calendar_days=("date", "count"),
            news_days=(news_col, lambda x: (x > 0).sum()),
            total_news_count=(news_col, "sum"),
            avg_news_count_all_days=(news_col, "mean")
        )
        .reset_index()
    )
    
    yearly["coverage_ratio"] = yearly["news_days"] / yearly["calendar_days"]
    yearly["category"] = cat
    
    yearly_coverage_list.append(yearly)

yearly_coverage = pd.concat(yearly_coverage_list, ignore_index=True)

yearly_coverage = yearly_coverage[
    ["category", "year", "calendar_days", "news_days",
     "coverage_ratio", "total_news_count", "avg_news_count_all_days"]
]

display(yearly_coverage)

,category,year,calendar_days,news_days,coverage_ratio,total_news_count,avg_news_count_all_days
0,demand,2007,145,12,0.082759,24,0.165517
1,demand,2008,366,46,0.125683,51,0.139344
2,demand,2009,365,37,0.101370,38,0.104110
3,demand,2010,365,49,0.134247,55,0.150685
4,demand,2011,365,46,0.126027,49,0.134247
...,...,...,...,...,...,...,...
75,financial,2022,365,299,0.819178,1047,2.868493
76,financial,2023,365,290,0.794521,1065,2.917808
77,financial,2024,366,309,0.844262,1351,3.691257
78,financial,2025,365,315,0.863014,1762,4.827397


In [10]:
# 연도별 커버리지 pivot
yearly_coverage_pivot = yearly_coverage.pivot(
    index="year",
    columns="category",
    values="coverage_ratio"
)

display(yearly_coverage_pivot)

category,demand,financial,geopolitical,supply
year,,,,
2007,0.082759,0.124138,0.289655,0.055172
2008,0.125683,0.202186,0.330601,0.090164
2009,0.101370,0.145205,0.345205,0.082192
2010,0.134247,0.205479,0.397260,0.106849
2011,0.126027,0.243836,0.410959,0.134247
2012,0.177596,0.278689,0.584699,0.177596
2013,0.150685,0.252055,0.578082,0.120548
2014,0.142466,0.224658,0.605479,0.150685
2015,0.197260,0.315068,0.616438,0.156164


In [11]:
# 커버리지 0인 연도 확인
zero_coverage_years = yearly_coverage[
    yearly_coverage["coverage_ratio"] == 0
].sort_values(["category", "year"])

display(zero_coverage_years.head(10))

# 커버리지 낮은 연도 확인
low_coverage_threshold = 0.20

low_coverage_years = yearly_coverage[
    (yearly_coverage["coverage_ratio"] > 0) &
    (yearly_coverage["coverage_ratio"] < low_coverage_threshold)
].sort_values(["category", "year"])

display(low_coverage_years.head(10))

,category,year,calendar_days,news_days,coverage_ratio,total_news_count,avg_news_count_all_days


,category,year,calendar_days,news_days,coverage_ratio,total_news_count,avg_news_count_all_days
0,demand,2007,145,12,0.082759,24,0.165517
1,demand,2008,366,46,0.125683,51,0.139344
2,demand,2009,365,37,0.101370,38,0.104110
3,demand,2010,365,49,0.134247,55,0.150685
4,demand,2011,365,46,0.126027,49,0.134247
5,demand,2012,366,65,0.177596,75,0.204918
6,demand,2013,365,55,0.150685,62,0.169863
7,demand,2014,365,52,0.142466,57,0.156164
8,demand,2015,365,72,0.197260,85,0.232877
9,demand,2016,366,68,0.185792,76,0.207650


[해석]

1.  커버리지가 0인 연도: 0
2.  20% 미만의 낮은 커버리지를 보이는 연도: 
- `demand` - 2007~2016년
- `supply` - 2007~2016년
- `financial` - 2007년, 2009년
- `geopolitical` - 0

=> 초기 구간의 감성 피처는 관측 밀도가 낮을 수 있으므로, 이후 7일 평균 생성 및 모델 해석 시 주의가 필요하다.

In [12]:
# 주별 커버리지
weekly_coverage_list = []

for cat, df in calendar_sentiment_dfs.items():
    tmp = df.copy()
    news_col = f"news_count_{cat}"
    
    # 주 시작일 계산(월요일 기준)
    tmp["week_start"] = tmp["date"] - pd.to_timedelta(tmp["date"].dt.weekday, unit="D")
    
    weekly = (
        tmp.groupby("week_start")
        .agg(
            calendar_days=("date", "count"),
            news_days=(news_col, lambda x: (x > 0).sum()), # 해당 주에 뉴스가 있는 날짜 수
            total_news_count=(news_col, "sum"), # 해당 주의 전체 기사 수
            avg_news_count_all_days=(news_col, "mean")
        )
        .reset_index()
    )
    
    weekly["coverage_ratio"] = weekly["news_days"] / weekly["calendar_days"]
    weekly["category"] = cat
    
    weekly_coverage_list.append(weekly)

weekly_coverage = pd.concat(weekly_coverage_list, ignore_index=True)

weekly_coverage = weekly_coverage[
    ["category", "week_start", "calendar_days", "news_days",
     "coverage_ratio", "total_news_count", "avg_news_count_all_days"]
]

In [13]:
# 주별 커버리지 pivot
weekly_coverage_pivot = weekly_coverage.pivot(
    index="week_start",
    columns="category",
    values="coverage_ratio"
)

display(weekly_coverage_pivot.head(20))

category,demand,financial,geopolitical,supply
week_start,,,,
2007-08-06,0.250000,0.250000,0.250000,0.250000
2007-08-13,0.142857,0.142857,0.428571,0.142857
2007-08-20,0.000000,0.142857,0.142857,0.000000
2007-08-27,0.000000,0.000000,0.000000,0.000000
2007-09-03,0.142857,0.142857,0.571429,0.000000
2007-09-10,0.000000,0.000000,0.285714,0.000000
2007-09-17,0.000000,0.142857,0.142857,0.142857
2007-09-24,0.000000,0.000000,0.000000,0.000000
2007-10-01,0.000000,0.142857,0.285714,0.000000


In [14]:
# 커버리지 0인 주 확인
zero_coverage_weeks = weekly_coverage[
    weekly_coverage["coverage_ratio"] == 0
].sort_values(["category", "week_start"])

display(zero_coverage_weeks.head(10))

# 커버리지 낮은 주 확인
low_weekly_coverage_threshold = 0.20

low_coverage_weeks = weekly_coverage[
    (weekly_coverage["coverage_ratio"] > 0) &
    (weekly_coverage["coverage_ratio"] < low_weekly_coverage_threshold)
].sort_values(["category", "week_start"])

display(low_coverage_weeks.head(10))

,category,week_start,calendar_days,news_days,coverage_ratio,total_news_count,avg_news_count_all_days
2,demand,2007-08-20,7,0,0.0,0,0.0
3,demand,2007-08-27,7,0,0.0,0,0.0
5,demand,2007-09-10,7,0,0.0,0,0.0
6,demand,2007-09-17,7,0,0.0,0,0.0
7,demand,2007-09-24,7,0,0.0,0,0.0
8,demand,2007-10-01,7,0,0.0,0,0.0
9,demand,2007-10-08,7,0,0.0,0,0.0
10,demand,2007-10-15,7,0,0.0,0,0.0
13,demand,2007-11-05,7,0,0.0,0,0.0
15,demand,2007-11-19,7,0,0.0,0,0.0


,category,week_start,calendar_days,news_days,coverage_ratio,total_news_count,avg_news_count_all_days
1,demand,2007-08-13,7,1,0.142857,1,0.142857
4,demand,2007-09-03,7,1,0.142857,1,0.142857
11,demand,2007-10-22,7,1,0.142857,1,0.142857
14,demand,2007-11-12,7,1,0.142857,1,0.142857
16,demand,2007-11-26,7,1,0.142857,1,0.142857
24,demand,2008-01-21,7,1,0.142857,2,0.285714
25,demand,2008-01-28,7,1,0.142857,1,0.142857
35,demand,2008-04-07,7,1,0.142857,1,0.142857
41,demand,2008-05-19,7,1,0.142857,1,0.142857
42,demand,2008-05-26,7,1,0.142857,1,0.142857


In [15]:
# 주별 커버리지 요약
weekly_coverage_summary = (
    weekly_coverage
    .groupby("category")
    .agg(
        total_weeks=("week_start", "count"),
        zero_news_weeks=("news_days", lambda x: (x == 0).sum()),
        one_news_day_weeks=("news_days", lambda x: (x == 1).sum()),
        low_coverage_weeks=("coverage_ratio", lambda x: ((x > 0) & (x < low_weekly_coverage_threshold)).sum()),
        avg_weekly_coverage=("coverage_ratio", "mean"),
        min_weekly_coverage=("coverage_ratio", "min"),
        max_weekly_coverage=("coverage_ratio", "max")
    )
    .reset_index()
)

display(weekly_coverage_summary)

,category,total_weeks,zero_news_weeks,one_news_day_weeks,low_coverage_weeks,avg_weekly_coverage,min_weekly_coverage,max_weekly_coverage
0,demand,974,204,211,210,0.372213,0.0,1.0
1,financial,974,87,192,191,0.440415,0.0,1.0
2,geopolitical,974,14,41,40,0.654994,0.0,1.0
3,supply,974,220,244,243,0.363413,0.0,1.0


[해석]

- 주별 커버리지 확인 결과, 뉴스가 전혀 없는 주가 존재한다.
- `supply`와 `demand`는 뉴스가 없는 주와 낮은 커버리지 주가 상대적으로 많다.
- 반면 `geopolitical`은 주별 커버리지가 가장 안정적이다.
- 따라서 이후 감성 피처 생성 시 7일 평균 감성과 함께 `news_count_7d`를 같이 사용한다.

#### 5-3. 뉴스 없는 연속 구간 확인

In [16]:
no_news_streaks = []

for cat, df in calendar_sentiment_dfs.items():
    tmp = df[["date", f"news_count_{cat}"]].copy()
    news_col = f"news_count_{cat}"
    
    tmp["no_news"] = (tmp[news_col] == 0).astype(int)
    
    # no_news 값이 바뀌는 지점마다 그룹 생성
    tmp["streak_group"] = (tmp["no_news"] != tmp["no_news"].shift()).cumsum()
    
    streak = (
        tmp[tmp["no_news"] == 1]
        .groupby("streak_group")
        .agg(
            start_date=("date", "min"),
            end_date=("date", "max"),
            length_days=("date", "count")
        )
        .reset_index(drop=True)
    )
    
    streak["category"] = cat
    no_news_streaks.append(streak)

no_news_streaks = pd.concat(no_news_streaks, ignore_index=True)

no_news_streaks = no_news_streaks[
    ["category", "start_date", "end_date", "length_days"]
].sort_values(["category", "length_days"], ascending=[True, False])

top5_no_news_streaks = (
    no_news_streaks
    .groupby("category")
    .head(5)
    .reset_index(drop=True)
)

display(top5_no_news_streaks)

,category,start_date,end_date,length_days
0,demand,2016-12-06,2017-01-23,49
1,demand,2007-09-10,2007-10-21,42
2,demand,2014-05-22,2014-07-02,42
3,demand,2008-02-28,2008-04-08,41
4,demand,2011-12-24,2012-01-31,39
5,financial,2009-04-03,2009-05-05,33
6,financial,2008-03-11,2008-04-07,28
7,financial,2008-12-20,2009-01-12,24
8,financial,2007-12-15,2008-01-06,23
9,financial,2014-01-09,2014-01-30,22


[해석]

<최대 뉴스 공백 기간>
- `supply`: 55일
- `demand`: 49일
- `financial`: 33일
- `geopolitical`: 23일

=> 뉴스가 없는 기간이 단발적으로만 발생한 것이 아니라, 여러 날 연속되는 구간이 존재한다.

=> 따라서 이후 7일 평균 감성 생성 시 뉴스 공백을 보완하기 위해 `news_count_7d`와 `no_news_7d`를 함께 고려한다.

#### 5-4. 일별 기사 수 분포

In [17]:
news_count_distribution = []

for cat, df in calendar_sentiment_dfs.items():
    news_col = f"news_count_{cat}"
    
    news_count_all = df[news_col]
    news_count_news_days = df.loc[df[news_col] > 0, news_col]
    
    news_count_distribution.append({
        "category": cat,
        "mean_all_days": news_count_all.mean(),
        "median_all_days": news_count_all.median(),
        "q75_all_days": news_count_all.quantile(0.75),
        "news_days": news_count_news_days.count(),
        "mean_news_days": news_count_news_days.mean(),
        "median_news_days": news_count_news_days.median(),
        "q75_news_days": news_count_news_days.quantile(0.75),
        "max_news_days": news_count_news_days.max()
    })

news_count_distribution = pd.DataFrame(news_count_distribution)

display(news_count_distribution)

,category,mean_all_days,median_all_days,q75_all_days,news_days,mean_news_days,median_news_days,q75_news_days,max_news_days
0,demand,1.186344,0.0,1.0,2532,3.190758,2.0,4.0,19
1,supply,1.310573,0.0,1.0,2472,3.610437,2.0,5.0,23
2,geopolitical,2.460793,1.0,3.0,4460,3.757399,2.0,5.0,52
3,financial,1.228634,0.0,1.0,2997,2.791792,2.0,4.0,20


[해석]

- 전체 날짜 기준 중앙값은 대부분 0으로, 뉴스가 없는 날이 많다.
- 따라서 이후 감성 피처 생성 시 일별 감성만 사용하기보다 7일 평균 감성과 `news_count`를 함께 고려한다.

## 6. 평균 감성 피처 생성

#### 6-1. 직전 7일 평균 감성 피처 조합별 테스트


| 항목 | 내용 |
|---|---|
| 기준 기간 | 각 날짜 기준 당일을 제외한 직전 7캘린더일(t-7 ~ t-1) |
| 감성 처리 | 뉴스 없는 날짜의 `sentiment`는 NaN 유지 |
| 평균 방식 | 단순 평균, 기사 수 가중 평균 |
| 비교 기준 | `min_periods = 1, 3, 5` |
| 선택 기준 | 결측 비율과 감성값 안정성 |
| 함께 생성할 피처 | news_{category}_7d, nonews_{category}_7d |

------------------------------------------------------------------------

| 평균 방식 | `min_periods` | 컬럼명 예시 |
|---|---:|---|
| 단순 평균 | 1 | `sent_demand_7d_m1` |
| 단순 평균 | 3 | `sent_demand_7d_m3` |
| 단순 평균 | 5 | `sent_demand_7d_m5` |
| 기사 수 가중 평균 | 1 | `sent_demand_7d_w1` |
| 기사 수 가중 평균 | 3 | `sent_demand_7d_w3` |
| 기사 수 가중 평균 | 5 | `sent_demand_7d_w5` |

In [18]:
# 직전 7일 평균 감성 피처 생성

min_periods_list = [1, 3, 5]
# 1: 직전 7일 중 감성값이 1개라도 있으면 평균 계산
# 3: 감성값이 최소 3개 있어야 평균 계산
# 5: 감성값이 최소 5개 있어야 평균 계산

sentiment_feature_dfs = {}

for cat, df in calendar_sentiment_dfs.items():
    tmp = df.sort_values("date").reset_index(drop=True).copy()
    # 날짜 정렬 후 rolling 적용
    
    sent_col = f"sentiment_{cat}"
    news_col = f"news_count_{cat}"
    
    # 당일 제외
    sent_shift = tmp[sent_col].shift(1)
    news_shift = tmp[news_col].shift(1)
    
    # 직전 7일 기사 수 합계
    tmp[f"news_{cat}_7d"] = (
        news_shift
        .rolling(window=7, min_periods=1)
        .sum()
        .fillna(0)
    )
    
    # 직전 7일 뉴스 없음 여부
    tmp[f"nonews_{cat}_7d"] = (tmp[f"news_{cat}_7d"] == 0).astype(int)

    sent_filled = sent_shift.fillna(0)
    news_eff = news_shift.where(sent_shift.notna(),0)
    
    for mp in min_periods_list:
        # 단순 평균
        tmp[f"sent_{cat}_7d_m{mp}"] = (
            sent_shift
            .rolling(window=7, min_periods=mp)
            .mean()
        )

        # 기사 수 가중 평균
        weighted_sum = (
            (sent_filled * news_eff)
            .rolling(window=7, min_periods=mp)
            .sum()
        )
        
        weight_sum = (
            news_eff
            .rolling(window=7, min_periods=mp)
            .sum()
        )
        
        tmp[f"sent_{cat}_7d_w{mp}"] = (
            weighted_sum / weight_sum.replace(0, np.nan)
            # weight_sum == 0이면 나누기 전에 NaN 처리
        ).replace([np.inf, -np.inf], np.nan)
        # inf, -inf가 혹시 생겨도 최종적으로 NaN 처리
        # 직전 7일 sum(감성점수 × 기사 수) / 직전 7일 sum(기사 수)
    
    sentiment_feature_dfs[cat] = tmp
    
    print(f"[{cat}] 7d sentiment features created")
    print(f"  shape: {tmp.shape}")
    print()

[demand] 7d sentiment features created
  shape: (6810, 16)

[supply] 7d sentiment features created
  shape: (6810, 16)

[geopolitical] 7d sentiment features created
  shape: (6810, 16)

[financial] 7d sentiment features created
  shape: (6810, 16)



In [19]:
# 생성된 피처 확인
for cat, df in sentiment_feature_dfs.items():
    feature_cols = [
        "date",
        f"sentiment_{cat}",
        f"news_count_{cat}", # 원본 일별 기사 수
        f"news_{cat}_7d",
        f"nonews_{cat}_7d",
        f"sent_{cat}_7d_m1",
        f"sent_{cat}_7d_m3",
        f"sent_{cat}_7d_m5",
        f"sent_{cat}_7d_w1",
        f"sent_{cat}_7d_w3",
        f"sent_{cat}_7d_w5",
    ]
    
    print(f"\n=== [{cat}] 7d feature sample ===")
    display(df[feature_cols].head(5))


=== [demand] 7d feature sample ===


,date,sentiment_demand,news_count_demand,news_demand_7d,nonews_demand_7d,sent_demand_7d_m1,sent_demand_7d_m3,sent_demand_7d_m5,sent_demand_7d_w1,sent_demand_7d_w3,sent_demand_7d_w5
0,2007-08-09,0.221255,13,0.0,1,NaN,NaN,NaN,NaN,NaN,NaN
1,2007-08-10,NaN,0,13.0,0,0.221255,NaN,NaN,0.221255,NaN,NaN
2,2007-08-11,NaN,0,13.0,0,0.221255,NaN,NaN,0.221255,0.221255,NaN
3,2007-08-12,NaN,0,13.0,0,0.221255,NaN,NaN,0.221255,0.221255,NaN
4,2007-08-13,NaN,0,13.0,0,0.221255,NaN,NaN,0.221255,0.221255,0.221255



=== [supply] 7d feature sample ===


,date,sentiment_supply,news_count_supply,news_supply_7d,nonews_supply_7d,sent_supply_7d_m1,sent_supply_7d_m3,sent_supply_7d_m5,sent_supply_7d_w1,sent_supply_7d_w3,sent_supply_7d_w5
0,2007-08-09,-0.01186,6,0.0,1,NaN,NaN,NaN,NaN,NaN,NaN
1,2007-08-10,NaN,0,6.0,0,-0.01186,NaN,NaN,-0.01186,NaN,NaN
2,2007-08-11,NaN,0,6.0,0,-0.01186,NaN,NaN,-0.01186,-0.01186,NaN
3,2007-08-12,NaN,0,6.0,0,-0.01186,NaN,NaN,-0.01186,-0.01186,NaN
4,2007-08-13,NaN,0,6.0,0,-0.01186,NaN,NaN,-0.01186,-0.01186,-0.01186



=== [geopolitical] 7d feature sample ===


,date,sentiment_geopolitical,news_count_geopolitical,news_geopolitical_7d,nonews_geopolitical_7d,sent_geopolitical_7d_m1,sent_geopolitical_7d_m3,sent_geopolitical_7d_m5,sent_geopolitical_7d_w1,sent_geopolitical_7d_w3,sent_geopolitical_7d_w5
0,2007-08-09,-0.066130,52,0.0,1,NaN,NaN,NaN,NaN,NaN,NaN
1,2007-08-10,NaN,0,52.0,0,-0.06613,NaN,NaN,-0.06613,NaN,NaN
2,2007-08-11,NaN,0,52.0,0,-0.06613,NaN,NaN,-0.06613,-0.06613,NaN
3,2007-08-12,NaN,0,52.0,0,-0.06613,NaN,NaN,-0.06613,-0.06613,NaN
4,2007-08-13,0.048347,1,52.0,0,-0.06613,NaN,NaN,-0.06613,-0.06613,-0.06613



=== [financial] 7d feature sample ===


,date,sentiment_financial,news_count_financial,news_financial_7d,nonews_financial_7d,sent_financial_7d_m1,sent_financial_7d_m3,sent_financial_7d_m5,sent_financial_7d_w1,sent_financial_7d_w3,sent_financial_7d_w5
0,2007-08-09,-0.138557,20,0.0,1,NaN,NaN,NaN,NaN,NaN,NaN
1,2007-08-10,NaN,0,20.0,0,-0.138557,NaN,NaN,-0.138557,NaN,NaN
2,2007-08-11,NaN,0,20.0,0,-0.138557,NaN,NaN,-0.138557,-0.138557,NaN
3,2007-08-12,NaN,0,20.0,0,-0.138557,NaN,NaN,-0.138557,-0.138557,NaN
4,2007-08-13,NaN,0,20.0,0,-0.138557,NaN,NaN,-0.138557,-0.138557,-0.138557


#### 6-2. 7일 평균 감성 조합별 결측치 및 안정성 비교

[확인 기준]

1. 결측치 비율

2. 감성값 안정성

3. 단순 평균과 가중 평균의 차이


In [20]:
sentiment_7d_compare = []

for cat, df in sentiment_feature_dfs.items():
    feature_cols = [
        f"sent_{cat}_7d_m1",
        f"sent_{cat}_7d_m3",
        f"sent_{cat}_7d_m5",

        f"sent_{cat}_7d_w1",
        f"sent_{cat}_7d_w3",
        f"sent_{cat}_7d_w5",
    ]
    
    for col in feature_cols:
        valid = df[col].dropna()
        
        sentiment_7d_compare.append({
            "category": cat,
            "feature": col,
            "missing_ratio": df[col].isna().mean(), # 결측치 비율
            "valid_count": valid.count(), # 결측치가 아닌 값의 개수
            "std": valid.std(), # 결측치를 제외한 감성 피처의 표준편차 -> 안정성 확인
            "q05": valid.quantile(0.05), # 5% 분위수 -> 극단값 확인
            "q95": valid.quantile(0.95), # 95% 분위수 -> 극단값 확인
        })

sentiment_7d_compare = pd.DataFrame(sentiment_7d_compare)

display(sentiment_7d_compare)

,category,feature,missing_ratio,valid_count,std,q05,q95
0,demand,sent_demand_7d_m1,0.203377,5425,0.447696,-0.894079,0.818964
1,demand,sent_demand_7d_m3,0.602056,2710,0.251178,-0.484594,0.336877
2,demand,sent_demand_7d_m5,0.722761,1888,0.198374,-0.352792,0.315609
3,demand,sent_demand_7d_w1,0.203377,5425,0.441012,-0.894079,0.818964
4,demand,sent_demand_7d_w3,0.203524,5424,0.441038,-0.894079,0.819068
5,demand,sent_demand_7d_w5,0.203818,5422,0.441090,-0.894079,0.819276
6,supply,sent_supply_7d_m1,0.222173,5297,0.429508,-0.914568,0.720759
7,supply,sent_supply_7d_m3,0.616740,2610,0.233356,-0.471589,0.291603
8,supply,sent_supply_7d_m5,0.729515,1842,0.177368,-0.337640,0.249829
9,supply,sent_supply_7d_w1,0.222173,5297,0.424295,-0.914568,0.722269


In [21]:
# 조합별 평균 비교 요약
sentiment_7d_compare["method"] = sentiment_7d_compare["feature"].str.extract(r"7d_([mw]\d)$")

method_summary = (
    sentiment_7d_compare
    .groupby("method")
    .agg(
        avg_missing_ratio=("missing_ratio", "mean"), # 조합별 평균 결측률
        max_missing_ratio=("missing_ratio", "max"), # 조합별 최대 결측률
        avg_std=("std", "mean"), # 조합별 평균 표준편차
        max_std=("std", "max"), # 조합별 최대 표준편차
        avg_valid_count=("valid_count", "mean") # 조합별 평균 유효 관측치 수
    ) 
    .reset_index()
    .sort_values("avg_missing_ratio")
)

display(method_summary)

,method,avg_missing_ratio,max_missing_ratio,avg_std,max_std,avg_valid_count
0,m1,0.132085,0.222173,0.379205,0.447696,5910.5
3,w1,0.132085,0.222173,0.370846,0.441012,5910.5
4,w3,0.132232,0.222320,0.370874,0.441038,5909.5
5,w5,0.132526,0.222614,0.370930,0.441090,5907.5
1,m3,0.460206,0.616740,0.241538,0.261365,3676.0
2,m5,0.653451,0.729515,0.190102,0.199389,2360.0


In [66]:
# 단순 평균 vs 가중 평균 상관 분석
mean_weight_corr = []

for cat, df in sentiment_feature_dfs.items():
    for mp in [1, 3, 5]:
        m_col = f"sent_{cat}_7d_m{mp}"
        w_col = f"sent_{cat}_7d_w{mp}"
        
        corr = df[[m_col, w_col]].corr().iloc[0, 1]
        
        mean_weight_corr.append({
            "category": cat,
            "min_periods": mp,
            "mean_feature": m_col,
            "weighted_feature": w_col,
            "corr": corr
        })

mean_weight_corr = pd.DataFrame(mean_weight_corr)

display(mean_weight_corr)

,category,min_periods,mean_feature,weighted_feature,corr
0,demand,1,sent_demand_7d_m1,sent_demand_7d_w1,0.987076
1,demand,3,sent_demand_7d_m3,sent_demand_7d_w3,0.929968
2,demand,5,sent_demand_7d_m5,sent_demand_7d_w5,0.866553
3,supply,1,sent_supply_7d_m1,sent_supply_7d_w1,0.986549
4,supply,3,sent_supply_7d_m3,sent_supply_7d_w3,0.916141
5,supply,5,sent_supply_7d_m5,sent_supply_7d_w5,0.837979
6,geopolitical,1,sent_geopolitical_7d_m1,sent_geopolitical_7d_w1,0.944259
7,geopolitical,3,sent_geopolitical_7d_m3,sent_geopolitical_7d_w3,0.923818
8,geopolitical,5,sent_geopolitical_7d_m5,sent_geopolitical_7d_w5,0.896601
9,financial,1,sent_financial_7d_m1,sent_financial_7d_w1,0.980326


[해석]

1. min_periods 비교
- `min_periods=1`: 평균 결측률이 약 13.2%로 가장 낮고, 유효 관측치 수가 가장 많다.
- `min_periods=3`: 평균 결측률이 약 46.0%로 증가하여 데이터 손실이 크다.
- `min_periods=5`: 평균 결측률이 약 65.3%까지 증가하여 최종 피처로 사용하기 어렵다.

=> 최종적으로 결측률과 활용 가능한 관측치 수를 고려하여 `min_periods=1`을 사용한다.

2. 단순 평균 vs 기사 수 가중 평균
- 단순 평균(`m1`): 기사 수를 별도 가중치로 반영하지 않고, 직전 7일 감성값을 단순 평균한 방식
- 기사 수 가중 평균(`w1`): 기사 수를 가중치로 반영한 방식

=> 단순 평균과의 상관계수가 모든 카테고리에서 0.94 이상으로 높게 나타났다.
=> 최종적으로 더 단순하고 해석이 쉬운 단순 평균을 사용한다.

## 7. 최종 감성 피처 데이터셋 생성

In [22]:
final_sentiment_features = calendar_df.copy()

for cat, df in sentiment_feature_dfs.items():
    selected = df[[
        "date",
        f"sent_{cat}_7d_m1",
        f"news_{cat}_7d",
        f"nonews_{cat}_7d"
    ]].copy()
    
    selected = selected.rename(columns={
        f"sent_{cat}_7d_m1": f"sent_{cat}_7d"
    })
    
    final_sentiment_features = final_sentiment_features.merge(
        selected,
        on="date",
        how="left"
    )

display(final_sentiment_features.head())
display(final_sentiment_features.tail())

print(f"shape: {final_sentiment_features.shape}")
print(f"date range: {final_sentiment_features['date'].min().date()} ~ {final_sentiment_features['date'].max().date()}")

,date,sent_demand_7d,news_demand_7d,nonews_demand_7d,sent_supply_7d,news_supply_7d,nonews_supply_7d,sent_geopolitical_7d,news_geopolitical_7d,nonews_geopolitical_7d,sent_financial_7d,news_financial_7d,nonews_financial_7d
0,2007-08-09,NaN,0.0,1,NaN,0.0,1,NaN,0.0,1,NaN,0.0,1
1,2007-08-10,0.221255,13.0,0,-0.01186,6.0,0,-0.06613,52.0,0,-0.138557,20.0,0
2,2007-08-11,0.221255,13.0,0,-0.01186,6.0,0,-0.06613,52.0,0,-0.138557,20.0,0
3,2007-08-12,0.221255,13.0,0,-0.01186,6.0,0,-0.06613,52.0,0,-0.138557,20.0,0
4,2007-08-13,0.221255,13.0,0,-0.01186,6.0,0,-0.06613,52.0,0,-0.138557,20.0,0


,date,sent_demand_7d,news_demand_7d,nonews_demand_7d,sent_supply_7d,news_supply_7d,nonews_supply_7d,sent_geopolitical_7d,news_geopolitical_7d,nonews_geopolitical_7d,sent_financial_7d,news_financial_7d,nonews_financial_7d
6805,2026-03-27,-0.128609,52.0,0,-0.228180,65.0,0,-0.203443,94.0,0,-0.116332,45.0,0
6806,2026-03-28,-0.152080,54.0,0,-0.278535,64.0,0,-0.237217,89.0,0,-0.160258,48.0,0
6807,2026-03-29,-0.116078,52.0,0,-0.211141,65.0,0,-0.156130,88.0,0,-0.160258,48.0,0
6808,2026-03-30,-0.305626,51.0,0,-0.090959,64.0,0,-0.364079,82.0,0,-0.198491,44.0,0
6809,2026-03-31,-0.244848,48.0,0,0.003237,52.0,0,-0.310264,80.0,0,-0.171467,42.0,0


shape: (6810, 13)
date range: 2007-08-09 ~ 2026-03-31


In [23]:
# 결측 확인
display(final_sentiment_features.isna().sum())

# NA 처리
sentiment_final_model = final_sentiment_features.copy()

sent_cols = [
    col for col in sentiment_final_model.columns
    if col.startswith("sent_")
]

for col in sent_cols:
    na_count = sentiment_final_model[col].isna().sum()
    print(f"{col}: {na_count:,} NA -> 0")
    sentiment_final_model[col] = sentiment_final_model[col].fillna(0)

display(sentiment_final_model.isna().sum())

date                         0
sent_demand_7d            1385
news_demand_7d               0
nonews_demand_7d             0
sent_supply_7d            1513
news_supply_7d               0
nonews_supply_7d             0
sent_geopolitical_7d        91
news_geopolitical_7d         0
nonews_geopolitical_7d       0
sent_financial_7d          609
news_financial_7d            0
nonews_financial_7d          0
dtype: int64

sent_demand_7d: 1,385 NA -> 0
sent_supply_7d: 1,513 NA -> 0
sent_geopolitical_7d: 91 NA -> 0
sent_financial_7d: 609 NA -> 0


date                      0
sent_demand_7d            0
news_demand_7d            0
nonews_demand_7d          0
sent_supply_7d            0
news_supply_7d            0
nonews_supply_7d          0
sent_geopolitical_7d      0
news_geopolitical_7d      0
nonews_geopolitical_7d    0
sent_financial_7d         0
news_financial_7d         0
nonews_financial_7d       0
dtype: int64

In [24]:
# 저장
SENTIMENT_FEATURE_DIR = Path("../data/Sentiment_Final")
save_path = SENTIMENT_FEATURE_DIR / "sentiment_avg_7d_final.csv"

sentiment_final_model.to_csv(save_path, index=False, encoding="utf-8-sig")

print(f"saved: {save_path}")
print(f"shape: {sentiment_final_model.shape}")

saved: ..\data\Sentiment_Final\sentiment_avg_7d_final.csv
shape: (6810, 13)
